File: hw/CEDENOR/U3/HW30Observer&Strategy/Python/ObserverPython/controller/Stock.py
# Nahomi Cedeño, codecreators, DCCO - ESPE
class Stock:
    def __init__(self, symbol='', price=0.0):
        self._symbol = symbol
        self._price = price
        self._investors = []

    @property
    def symbol(self):
        return self._symbol

    @symbol.setter
    def symbol(self, symbol):
        self._symbol = symbol

    @property
    def price(self):
        return self._price

    @price.setter
    def price(self, price):
        self._price = price

    def add_observer(self, investor):
        self._investors.append(investor)

    def delete_observer(self, investor):
        self._investors.remove(investor)

    def notify_observers(self, args):
        for investor in self._investors:
            investor.update(self, args)

File: hw/CEDENOR/U3/HW30Observer&Strategy/Java/ObserverPatttern/src/main/java/ec/edu/espe/observerpattern/controller/Stock.java
package ec.edu.espe.observerpattern.controller;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 *
 * @author Nahomi Cedeño, codecreators, DCCO - ESPE
 */
abstract class Stock {
    protected String symbol;
    protected double price;
    private List<IInvestor> investors = new ArrayList<>();

    public Stock() {
     
    }
    
    public String getSymbol() {  
        return symbol;
    }

    public void addObserver(IInvestor investor) {
        investors.add(investor);
    }

    public void deleteObserver(IInvestor investor) {
        investors.remove(investor);
    }

    protected void notifyObservers(Object args) {
        Iterator i = investors.iterator();
        while (i.hasNext()){
            IInvestor investor = (IInvestor)i.next();
            investor.update(this, args);
        }
    }
}
